crates/cranelift/src/compiler/component.rs

```diff
@@ -0,0 +1,170 @@
+        let offsets = VMComponentOffsets::new(isa.pointer_bytes(), component);
+
+        let CompilerContext {
+            mut func_translator,
+            codegen_context: mut context,
+        } = self.take_context();
+
+        context.func = ir::Function::with_name_signature(
+            ir::ExternalName::user(0, 0),
+            crate::indirect_signature(isa, ty),
+        );
+
+        let mut builder = FunctionBuilder::new(&mut context.func, func_translator.context());
+        let block0 = builder.create_block();
+
+        // Start off by spilling all the wasm arguments into a stack slot to be
+        // passed to the host function.
+        let (values_vec_ptr_val, values_vec_len) =
+            self.wasm_to_host_spill_args(ty, &mut builder, block0);
+        let vmctx = builder.func.dfg.block_params(block0)[0];
+
+        // Below this will incrementally build both the signature of the host
+        // function we're calling as well as the list of arguments since the
+        // list is somewhat long.
+        let mut callee_args = Vec::new();
+        let mut host_sig = ir::Signature::new(crate::wasmtime_call_conv(isa));
+
+        let CanonicalOptions {
+            memory,
+            realloc,
+            string_encoding,
+        } = lowering.options;
+
+        // vmctx: *mut VMComponentContext
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(vmctx);
+
+        // data: *mut u8,
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(builder.ins().load(
+            pointer_type,
+            MemFlags::trusted(),
+            vmctx,
+            i32::try_from(offsets.lowering_data(lowering.index)).unwrap(),
+        ));
+
+        // memory: *mut VMMemoryDefinition
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(match memory {
+            Some(idx) => builder.ins().load(
+                pointer_type,
+                MemFlags::trusted(),
+                vmctx,
+                i32::try_from(offsets.runtime_memory(idx)).unwrap(),
+            ),
+            None => builder.ins().iconst(pointer_type, 0),
+        });
+
+        // realloc: *mut VMCallerCheckedAnyfunc
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(match realloc {
+            Some(idx) => builder.ins().load(
+                pointer_type,
+                MemFlags::trusted(),
+                vmctx,
+                i32::try_from(offsets.runtime_realloc(idx)).unwrap(),
+            ),
+            None => builder.ins().iconst(pointer_type, 0),
+        });
+
+        // string_encoding: StringEncoding
+        host_sig.params.push(ir::AbiParam::new(ir::types::I8));
+        callee_args.push(
+            builder
+                .ins()
+                .iconst(ir::types::I8, i64::from(string_encoding as u8)),
+        );
+
+        // storage: *mut ValRaw
```

Comment:
True yeah, although I sort of just invented the `ValRaw` name as it was the raw version of `wasmtime::Val`. We don't otherwise export `VM*` types to the outside world which `wasmtime::ValRaw` is, though. One way to solve this perhaps would be a `struct ValRaw(VMValRaw)` or similar, but I'm not sure if it's worth it per se.